src/uu/dd/src/dd.rs

```diff
@@ -0,0 +1,1296 @@
+            // debug_assert!(wstat.bytes_total >= n);
+            wstat.bytes_total <= n
+        }
+        None => true,
+    }
+}
+
+/// Perform the copy/convert operations. Stdout version
+// Note: Some of dd's functionality depends on whether the output is actually a file. This breaks the Output<Write> abstraction,
+// and should be fixed in the future.
+fn dd_stdout<R: Read>(mut i: Input<R>, mut o: Output<io::Stdout>) -> Result<(), Box<dyn Error>> {
+    let mut rstat = ReadStat {
+        reads_complete: 0,
+        reads_partial: 0,
+        records_truncated: 0,
+    };
+    let mut wstat = WriteStat {
+        writes_complete: 0,
+        writes_partial: 0,
+        bytes_total: 0,
+    };
+    let start = time::Instant::now();
+    let bsize = calc_bsize(i.ibs, o.obs);
+
+    let prog_tx = {
+        let (tx, rx) = mpsc::channel();
+        thread::spawn(gen_prog_updater(rx, i.xfer_stats));
+        tx
+    };
+
+    while below_count_limit(&i.count, &rstat, &wstat) {
+        // Read/Write
+        let loop_bsize = calc_loop_bsize(&i.count, &rstat, &wstat, i.ibs, bsize);
+        match read_helper(&mut i, loop_bsize)? {
+            (
+                ReadStat {
+                    reads_complete: 0,
+                    reads_partial: 0,
+                    ..
+                },
+                _,
+            ) => break,
+            (rstat_update, buf) => {
+                let wstat_update = o.write_blocks(buf)?;
+
+                rstat += rstat_update;
+                wstat += wstat_update;
+            }
+        };
+        // Update Prog
+        prog_tx.send(ProgUpdate {
+            reads_complete: rstat.reads_complete,
+            reads_partial: rstat.reads_partial,
+            writes_complete: wstat.writes_complete,
+            writes_partial: wstat.writes_partial,
+            bytes_total: wstat.bytes_total,
+            records_truncated: rstat.records_truncated,
+            duration: start.elapsed(),
+        })?;
+    }
+
+    if o.cflags.fsync {
+        o.fsync()?;
+    } else if o.cflags.fdatasync {
+        o.fdatasync()?;
+    }
+
+    match i.xfer_stats {
+        Some(StatusLevel::Noxfer) | Some(StatusLevel::None) => {}
+        _ => print_xfer_stats(&ProgUpdate {
+            reads_complete: rstat.reads_complete,
+            reads_partial: rstat.reads_partial,
+            writes_complete: wstat.writes_complete,
+            writes_partial: wstat.writes_partial,
+            bytes_total: wstat.bytes_total,
+            records_truncated: rstat.records_truncated,
+            duration: start.elapsed(),
+        }),
+    }
+    Ok(())
+}
+
+/// Perform the copy/convert operations. File backed output version
+// Note: Some of dd's functionality depends on whether the output is actually a file. This breaks the Output<Write> abstraction,
+// and should be fixed in the future.
+fn dd_fileout<R: Read>(mut i: Input<R>, mut o: Output<File>) -> Result<(), Box<dyn Error>> {
```

Comment:
I agree this is confusing, and I would love to fix it, but I don't currently know how. I ran into this problem early on in implementation, tried to fix it several times, but never quite got there. I've been doing some research over the past few days, and I think I can use OO features and dynamic dispatch to do what I want. I'm not really an OO programmer, so I skipped that chapter in the Rust book. I've added a comment for now, but this will probably be the fist issue I tackle when the code is merged. It's a major maintenance headache to keep dd_fileout the same as dd_stdout.

Reply:
We can use traits to solve this. I don't know what the best approach would be, but here's one possiblity:
<details>
<summary> Patch </summary>

```
diff --git a/src/uu/dd/src/dd.rs b/src/uu/dd/src/dd.rs
index 2b44a47d..5f7af9ea 100644
--- a/src/uu/dd/src/dd.rs
+++ b/src/uu/dd/src/dd.rs
@@ -289,13 +289,19 @@ impl<R: Read> Input<R> {
     }
 }
 
+trait OutputTrait: Sized + Write {
+    fn new(matches: &Matches) -> Result<Self, Box<dyn Error>>;
+    fn fsync(&mut self) -> io::Result<()>;
+    fn fdatasync(&mut self) -> io::Result<()>;
+}
+
 struct Output<W: Write> {
     dst: W,
     obs: usize,
     cflags: OConvFlags,
 }
 
-impl Output<io::Stdout> {
+impl OutputTrait for Output<io::Stdout> {
     fn new(matches: &Matches) -> Result<Self, Box<dyn Error>> {
         let obs = parseargs::parse_obs(matches)?;
         let cflags = parseargs::parse_conv_flag_output(matches)?;
@@ -314,6 +320,111 @@ impl Output<io::Stdout> {
     }
 }
 
+impl<W: Write> Output<W>
+where
+    Self: OutputTrait,
+{
+    fn write_blocks(&mut self, buf: Vec<u8>) -> io::Result<WriteStat> {
+        let mut writes_complete = 0;
+        let mut writes_partial = 0;
+        let mut base_idx = 0;
+
+        while base_idx < buf.len() {
+            let next_blk = cmp::min(base_idx + self.obs, buf.len());
+            let plen = next_blk - base_idx;
+
+            match self.write(&buf[base_idx..next_blk])? {
+                wlen if wlen < plen => {
+                    writes_partial += 1;
+                    base_idx += wlen;
+                }
+                wlen => {
+                    writes_complete += 1;
+                    base_idx += wlen;
+                }
+            }
+        }
+
+        Ok(WriteStat {
+            writes_complete,
+            writes_partial,
+            bytes_total: base_idx.try_into().unwrap_or(0u128),
+        })
+    }
+
+    fn dd_out<R: Read>(mut self, mut i: Input<R>) -> Result<(), Box<dyn Error>> {
+        let mut rstat = ReadStat {
+            reads_complete: 0,
+            reads_partial: 0,
+            records_truncated: 0,
+        };
+        let mut wstat = WriteStat {
+            writes_complete: 0,
+            writes_partial: 0,
+            bytes_total: 0,
+        };
+        let start = time::Instant::now();
+        let bsize = calc_bsize(i.ibs, self.obs);
+
+        let prog_tx = {
+            let (tx, rx) = mpsc::channel();
+            thread::spawn(gen_prog_updater(rx, i.xfer_stats));
+            tx
+        };
+
+        while below_count_limit(&i.count, &rstat, &wstat) {
+            // Read/Write
+            let loop_bsize = calc_loop_bsize(&i.count, &rstat, &wstat, i.ibs, bsize);
+            match read_helper(&mut i, loop_bsize)? {
+                (
+                    ReadStat {
+                        reads_complete: 0,
+                        reads_partial: 0,
+                        ..
+                    },
+                    _,
+                ) => break,
+                (rstat_update, buf) => {
+                    let wstat_update = self.write_blocks(buf)?;
+
+                    rstat += rstat_update;
+                    wstat += wstat_update;
+                }
+            };
+            // Update Prog
+            prog_tx.send(ProgUpdate {
+                reads_complete: rstat.reads_complete,
+                reads_partial: rstat.reads_partial,
+                writes_complete: wstat.writes_complete,
+                writes_partial: wstat.writes_partial,
+                bytes_total: wstat.bytes_total,
+                records_truncated: rstat.records_truncated,
+                duration: start.elapsed(),
+            })?;
+        }
+
+        if self.cflags.fsync {
+            self.fsync()?;
+        } else if self.cflags.fdatasync {
+            self.fdatasync()?;
+        }
+
+        match i.xfer_stats {
+            Some(StatusLevel::Noxfer) | Some(StatusLevel::None) => {}
+            _ => print_xfer_stats(&ProgUpdate {
+                reads_complete: rstat.reads_complete,
+                reads_partial: rstat.reads_partial,
+                writes_complete: wstat.writes_complete,
+                writes_partial: wstat.writes_partial,
+                bytes_total: wstat.bytes_total,
+                records_truncated: rstat.records_truncated,
+                duration: start.elapsed(),
+            }),
+        }
+        Ok(())
+    }
+}
+
 #[cfg(target_os = "linux")]
 fn make_linux_oflags(oflags: &OFlags) -> Option<libc::c_int> {
     let mut flag = 0;
@@ -354,7 +465,7 @@ fn make_linux_oflags(oflags: &OFlags) -> Option<libc::c_int> {
     }
 }
 
-impl Output<File> {
+impl OutputTrait for Output<File> {
     fn new(matches: &Matches) -> Result<Self, Box<dyn Error>> {
         fn open_dst(
             path: &Path,
@@ -450,66 +561,6 @@ impl Write for Output<io::Stdout> {
     }
 }
 
-impl Output<io::Stdout> {
-    fn write_blocks(&mut self, buf: Vec<u8>) -> io::Result<WriteStat> {
-        let mut writes_complete = 0;
-        let mut writes_partial = 0;
-        let mut base_idx = 0;
-
-        while base_idx < buf.len() {
-            let next_blk = cmp::min(base_idx + self.obs, buf.len());
-            let plen = next_blk - base_idx;
-
-            match self.write(&buf[base_idx..next_blk])? {
-                wlen if wlen < plen => {
-                    writes_partial += 1;
-                    base_idx += wlen;
-                }
-                wlen => {
-                    writes_complete += 1;
-                    base_idx += wlen;
-                }
-            }
-        }
-
-        Ok(WriteStat {
-            writes_complete,
-            writes_partial,
-            bytes_total: base_idx.try_into().unwrap_or(0u128),
-        })
-    }
-}
-
-impl Output<File> {
-    fn write_blocks(&mut self, buf: Vec<u8>) -> io::Result<WriteStat> {
-        let mut writes_complete = 0;
-        let mut writes_partial = 0;
-        let mut base_idx = 0;
-
-        while base_idx < buf.len() {
-            let next_blk = cmp::min(base_idx + self.obs, buf.len());
-            let plen = next_blk - base_idx;
-
-            match self.write(&buf[base_idx..next_blk])? {
-                wlen if wlen < plen => {
-                    writes_partial += 1;
-                    base_idx += wlen;
-                }
-                wlen => {
-                    writes_complete += 1;
-                    base_idx += wlen;
-                }
-            }
-        }
-
-        Ok(WriteStat {
-            writes_complete,
-            writes_partial,
-            bytes_total: base_idx.try_into().unwrap_or(0u128),
-        })
-    }
-}
-
 /// Splits the content of buf into cbs-length blocks
 /// Appends padding as specified by conv=block and cbs=N
 fn block(buf: Vec<u8>, cbs: usize, rstat: &mut ReadStat) -> Vec<Vec<u8>> {
@@ -890,156 +941,6 @@ fn below_count_limit(count: &Option<CountType>, rstat: &ReadStat, wstat: &WriteS
     }
 }
 
-/// Perform the copy/convert operations. Stdout version
-// Note: Some of dd's functionality depends on whether the output is actually a file. This breaks the Output<Write> abstraction,
-// and should be fixed in the future.
-fn dd_stdout<R: Read>(mut i: Input<R>, mut o: Output<io::Stdout>) -> Result<(), Box<dyn Error>> {
-    let mut rstat = ReadStat {
-        reads_complete: 0,
-        reads_partial: 0,
-        records_truncated: 0,
-    };
-    let mut wstat = WriteStat {
-        writes_complete: 0,
-        writes_partial: 0,
-        bytes_total: 0,
-    };
-    let start = time::Instant::now();
-    let bsize = calc_bsize(i.ibs, o.obs);
-
-    let prog_tx = {
-        let (tx, rx) = mpsc::channel();
-        thread::spawn(gen_prog_updater(rx, i.xfer_stats));
-        tx
-    };
-
-    while below_count_limit(&i.count, &rstat, &wstat) {
-        // Read/Write
-        let loop_bsize = calc_loop_bsize(&i.count, &rstat, &wstat, i.ibs, bsize);
-        match read_helper(&mut i, loop_bsize)? {
-            (
-                ReadStat {
-                    reads_complete: 0,
-                    reads_partial: 0,
-                    ..
-                },
-                _,
-            ) => break,
-            (rstat_update, buf) => {
-                let wstat_update = o.write_blocks(buf)?;
-
-                rstat += rstat_update;
-                wstat += wstat_update;
-            }
-        };
-        // Update Prog
-        prog_tx.send(ProgUpdate {
-            reads_complete: rstat.reads_complete,
-            reads_partial: rstat.reads_partial,
-            writes_complete: wstat.writes_complete,
-            writes_partial: wstat.writes_partial,
-            bytes_total: wstat.bytes_total,
-            records_truncated: rstat.records_truncated,
-            duration: start.elapsed(),
-        })?;
-    }
-
-    if o.cflags.fsync {
-        o.fsync()?;
-    } else if o.cflags.fdatasync {
-        o.fdatasync()?;
-    }
-
-    match i.xfer_stats {
-        Some(StatusLevel::Noxfer) | Some(StatusLevel::None) => {}
-        _ => print_xfer_stats(&ProgUpdate {
-            reads_complete: rstat.reads_complete,
-            reads_partial: rstat.reads_partial,
-            writes_complete: wstat.writes_complete,
-            writes_partial: wstat.writes_partial,
-            bytes_total: wstat.bytes_total,
-            records_truncated: rstat.records_truncated,
-            duration: start.elapsed(),
-        }),
-    }
-    Ok(())
-}
-
-/// Perform the copy/convert operations. File backed output version
-// Note: Some of dd's functionality depends on whether the output is actually a file. This breaks the Output<Write> abstraction,
-// and should be fixed in the future.
-fn dd_fileout<R: Read>(mut i: Input<R>, mut o: Output<File>) -> Result<(), Box<dyn Error>> {
-    let mut rstat = ReadStat {
-        reads_complete: 0,
-        reads_partial: 0,
-        records_truncated: 0,
-    };
-    let mut wstat = WriteStat {
-        writes_complete: 0,
-        writes_partial: 0,
-        bytes_total: 0,
-    };
-    let start = time::Instant::now();
-    let bsize = calc_bsize(i.ibs, o.obs);
-
-    let prog_tx = {
-        let (tx, rx) = mpsc::channel();
-        thread::spawn(gen_prog_updater(rx, i.xfer_stats));
-        tx
-    };
-
-    while below_count_limit(&i.count, &rstat, &wstat) {
-        // Read/Write
-        let loop_bsize = calc_loop_bsize(&i.count, &rstat, &wstat, i.ibs, bsize);
-        match read_helper(&mut i, loop_bsize)? {
-            (
-                ReadStat {
-                    reads_complete: 0,
-                    reads_partial: 0,
-                    ..
-                },
-                _,
-            ) => break,
-            (rstat_update, buf) => {
-                let wstat_update = o.write_blocks(buf)?;
-
-                rstat += rstat_update;
-                wstat += wstat_update;
-            }
-        };
-        // Update Prog
-        prog_tx.send(ProgUpdate {
-            reads_complete: rstat.reads_complete,
-            reads_partial: rstat.reads_partial,
-            writes_complete: wstat.writes_complete,
-            writes_partial: wstat.writes_partial,
-            bytes_total: wstat.bytes_total,
-            records_truncated: rstat.records_truncated,
-            duration: start.elapsed(),
-        })?;
-    }
-
-    if o.cflags.fsync {
-        o.fsync()?;
-    } else if o.cflags.fdatasync {
-        o.fdatasync()?;
-    }
-
-    match i.xfer_stats {
-        Some(StatusLevel::Noxfer) | Some(StatusLevel::None) => {}
-        _ => print_xfer_stats(&ProgUpdate {
-            reads_complete: rstat.reads_complete,
-            reads_partial: rstat.reads_partial,
-            writes_complete: wstat.writes_complete,
-            writes_partial: wstat.writes_partial,
-            bytes_total: wstat.bytes_total,
-            records_truncated: rstat.records_truncated,
-            duration: start.elapsed(),
-        }),
-    }
-    Ok(())
-}
-
 // The compiler does not like Clippy's suggestion to use &str in place of &String here.
 #[allow(clippy::ptr_arg)]
 fn append_dashes_if_not_present(mut acc: Vec<String>, s: &String) -> Vec<String> {
@@ -1091,7 +992,7 @@ pub fn uumain(args: impl uucore::Args) -> i32 {
             let (i, o) =
                 unpack_or_rtn!(Input::<File>::new(&matches), Output::<File>::new(&matches));
 
-            dd_fileout(i, o)
+            o.dd_out(i)
         }
         (false, true) => {
             let (i, o) = unpack_or_rtn!(
@@ -1099,7 +1000,7 @@ pub fn uumain(args: impl uucore::Args) -> i32 {
                 Output::<File>::new(&matches)
             );
 
-            dd_fileout(i, o)
+            o.dd_out(i)
         }
         (true, false) => {
             let (i, o) = unpack_or_rtn!(
@@ -1107,7 +1008,7 @@ pub fn uumain(args: impl uucore::Args) -> i32 {
                 Output::<io::Stdout>::new(&matches)
             );
 
-            dd_stdout(i, o)
+            o.dd_out(i)
         }
         (false, false) => {
             let (i, o) = unpack_or_rtn!(
@@ -1115,7 +1016,7 @@ pub fn uumain(args: impl uucore::Args) -> i32 {
                 Output::<io::Stdout>::new(&matches)
             );
 
-            dd_stdout(i, o)
+            o.dd_out(i)
         }
     };
     match result {
diff --git a/src/uu/dd/src/dd_unit_tests/conversion_tests.rs b/src/uu/dd/src/dd_unit_tests/conversion_tests.rs
index 738ead5f..fdaadb48 100644
--- a/src/uu/dd/src/dd_unit_tests/conversion_tests.rs
+++ b/src/uu/dd/src/dd_unit_tests/conversion_tests.rs
@@ -153,7 +153,7 @@ fn all_valid_ascii_ebcdic_ascii_roundtrip_conv_test() {
         cflags: DEFAULT_CFO,
     };
 
-    dd_fileout(i, o).unwrap();
+    o.dd_out(i).unwrap();
 
     // EBCDIC->ASCII
     let test_name = "all-valid-ebcdic-to-ascii";
@@ -175,7 +175,7 @@ fn all_valid_ascii_ebcdic_ascii_roundtrip_conv_test() {
         cflags: DEFAULT_CFO,
     };
 
-    dd_fileout(i, o).unwrap();
+    o.dd_out(i).unwrap();
 
     // Final Comparison
     let res = File::open(&tmp_fname_ea).unwrap();
diff --git a/src/uu/dd/src/dd_unit_tests/mod.rs b/src/uu/dd/src/dd_unit_tests/mod.rs
index 49edadd9..09ec5222 100644
--- a/src/uu/dd/src/dd_unit_tests/mod.rs
+++ b/src/uu/dd/src/dd_unit_tests/mod.rs
@@ -121,7 +121,7 @@ macro_rules! make_spec_test (
         #[test]
         fn $test_id()
         {
-            dd_fileout($i,$o).unwrap();
+            $o.dd_out($i).unwrap();
 
             let res = File::open($tmp_fname).unwrap();
             // Check test file isn't empty (unless spec file is too)
```
</details>

Reply:
From looking at the changes, this looks really promising. How do you (@miDeb) want to proceed with the change? I'm happy to get this PR merged then let you tackle this issue. Or is there a way for you to merge with this PR while it's in progress?

Reply:
I think it's ok to merge this with the code duplication, since it's not a bug but a code-quality thing. I would open a follow-up PR afterwards with this change.